### PR TITLE
Implement some new features to the sections app

### PR DESCRIPTION
**Pull request by dan-gamble**
Fields can be required on a per type basis
Adds fields that should be searched by watson
Can have help text on a per field, per type basis
Creates section type templates in their group folder

### CODE REVIEW

**Review by jamesfoley — APPROVED** [review]

File: {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/apps/sections/models.py [hunk 1]
```diff
@@ -159,7 +209,50 @@ class Meta:
...
+                        anchor = f'id_{self._meta.model_name}_set-{self.order}-{field}'
+                        fields_str += f'<a href="#{anchor}">{field.title()}</a>{connector}'
+
+                    fields['__all__'] = ValidationError(mark_safe(f"{fields_str} fields are required"), code='error')
+
+                    raise ValidationError(fields)
+
+        if self.link_text and (not self.link_page or not self.link_url):
+            raise ValidationError({
+                'link_page': 'Please supply 1 of "link page" or "link URL"',
```
Comment (danielsamuels) [inline {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/apps/sections/models.py]: I'd probably change this to be 

> Please provide either a "Link Page" or a "Link URL"

File: {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/apps/sections/models.py [hunk 2]
```diff
@@ -174,16 +267,43 @@ def link_location(self):
...
 
+    def get_searchable_text(self):
+        """Returns a blob of text suitable for searching."""
 
-class ContentSection(SectionBase):
+        # Let's look for the options for our section type.
+        for section_group in SECTION_TYPES:
+            for section_type in section_group[1]['sections']:
+                section_label = slugify(
+                    '{}-{}'.format(section_group[0], section_type[0]))
```
Comment (danielsamuels) [inline {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/apps/sections/models.py]: Switch to f-string?

File: {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/apps/sections/models.py [hunk 3]
```diff
@@ -174,16 +267,43 @@ def link_location(self):
...
+
+                search_fields = section_options['search_fields']
+
+                search_text_items = []
+                for field in search_fields:
+                    search_item = getattr(self, field)
+                    if search_item:
+                        search_text_items.append(strip_tags(search_item))
+
+                return u'\n'.join(search_text_items)
```
Comment (danielsamuels) [inline {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/apps/sections/models.py]: Is `u` still needed under Py3?

File: {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/apps/sections/templatetags/sections.py [hunk 4]
```diff
@@ -17,45 +16,22 @@ def render_section(context, page_section):
...
-            os.system(f'touch {{ cookiecutter.package_name }}/apps/sections/templates/sections/types/{page_section.template}')
+            try:
+                os.makedirs(f'example_project/apps/sections/templates/sections/types/{page_section.template["folder"]}')
+            except OSError as exc:  # Python >2.5
+                if exc.errno == errno.EEXIST and os.path.isdir(page_section.template["folder"]):
+                    pass
+                else:
+                    raise
+
+            os.system(f'touch example_project/apps/sections/templates/sections/types/{page_section.template["folder"]}/{page_section.template["file_name"]}')
```
Comment (danielsamuels) [inline {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/apps/sections/templatetags/sections.py]: `example_project` -> `{{ cookiecutter.package_name }}`

File: {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/apps/sections/templatetags/sections.py [hunk 5]
```diff
@@ -17,45 +16,22 @@ def render_section(context, page_section):
...
-        return render_to_string('sections/types/{}'.format(page_section.template), context)
+        return render_to_string(f'sections/types/{page_section.template["folder"]}/{page_section.template["file_name"]}', context)
     except TemplateDoesNotExist as e:
         import os
         from django.conf import settings
 
         if settings.DEBUG:
-            os.system(f'touch {{ cookiecutter.package_name }}/apps/sections/templates/sections/types/{page_section.template}')
+            try:
+                os.makedirs(f'example_project/apps/sections/templates/sections/types/{page_section.template["folder"]}')
```
Comment (danielsamuels) [inline {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/apps/sections/templatetags/sections.py]: `example_project`

File: {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/apps/sections/models.py [hunk 6]
```diff
@@ -159,7 +209,50 @@ class Meta:
...
+
+                        if index == fields_len - 2:
+                            connector = ' and '
+                        elif index == fields_len - 1:
+                            connector = ''
+
+                        anchor = f'id_{self._meta.model_name}_set-{self.order}-{field}'
+                        fields_str += f'<a href="#{anchor}">{field.title()}</a>{connector}'
+
+                    fields['__all__'] = ValidationError(mark_safe(f"{fields_str} fields are required"), code='error')
```
Comment (danielsamuels) [inline {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/apps/sections/models.py]: Switch to single quotes.

**Review by danielsamuels — CHANGES_REQUESTED** [review]

**Review by danielsamuels — APPROVED** [review]

**Pull request merged**